include/deal.II/particles/property_pool.h

```diff
@@ -115,14 +124,37 @@ namespace Particles
      */
     const unsigned int n_properties;
 
+    /**
+     * The currently allocated properties (whether assigned to
+     * a particle or available for assignment).
+     */
+    std::vector<double> properties;
+
     /**
      * A collection of handles that have been created by
-     * allocate_properties_array() and have not been destroyed by
-     * deallocate_properties_array().
+     * allocate_properties_array() and have been destroyed by
+     * deallocate_properties_array(). Since the memory is still
+     * allocated these handles can be reused for new particles
+     * to avoid memory allocation.
      */
-    std::set<Handle> currently_open_handles;
+    std::vector<Handle> currently_available_handles;
```

Comment:
The data structure you're looking for is one where you can cheaply `push_back()` and `pop_back()`. You don't actually need random access, and you don't need to access anything other than the top element I believe. So that would call for `std::stack`.

By default, `std::stack` builds on `std::deque`, which has the kind of storage allocation behavior you want: https://en.cppreference.com/w/cpp/container/deque

Reply:
I don't think it is critical here but it could be nice to compare it to the current version with @gassmoeller's benchmark setup.

Reply:
I think the advantage of `std::stack` is that when you come to the end of a `ParticleHandler`, you're going to put *all* of the handles into this variable, one after the other -- so it's going to grow a lot and you will have to do many re-allocations and copies. `std::stack`/`std::deque` don't require these re-allocations and copies because they are generally implemented as a linked list of modestly sized arrays.

Reply:
Does it matter? Who knows -- but it's easy to stay on the safe side :-)

Reply:
In principle I agree with your description, but there are two reasons I would like to stick to the vector for now: 1. The timings in my benchmark case are nearly identical, even deallocating 1e7 entries is nearly identical between vector and stack for me. Vector:
```
+---------------------------------------------+------------+------------+
| Total wallclock time elapsed since start    |      1.68s |            |
|                                             |            |            |
| Section                         | no. calls |  wall time | % of total |
+---------------------------------+-----------+------------+------------+
| Property allocation             |         1 |      1.01s |        60% |
| Property deallocation 1, 25%    |         1 |    0.0862s |       5.1% |
| Property deallocation 2, 25%    |         1 |    0.0891s |       5.3% |
| Property deallocation, all      |         1 |     0.203s |        12% |
| Property iteration, sorted      |         1 |     0.048s |       2.9% |
| Property iteration, unsorted    |         1 |    0.0898s |       5.3% |
| Property reallocation, 50%   |         1 |    0.0699s |       4.2% |
+---------------------------------+-----------+------------+------------+
```
Stack:
``` 
+---------------------------------------------+------------+------------+
| Total wallclock time elapsed since start    |      1.81s |            |
|                                             |            |            |
| Section                         | no. calls |  wall time | % of total |
+---------------------------------+-----------+------------+------------+
| Property allocation             |         1 |      1.07s |        59% |
| Property deallocation 1, 25%    |         1 |    0.0819s |       4.5% |
| Property deallocation 2, 25%    |         1 |    0.0785s |       4.3% |
| Property deallocation, all      |         1 |      0.26s |        14% |
| Property iteration, sorted      |         1 |    0.0474s |       2.6% |
| Property iteration, unsorted    |         1 |    0.0895s |         5% |
| Property reallocation, 50%   |         1 |     0.106s |       5.9% |
+---------------------------------+-----------+------------+------------+
```

Maybe my stack is a vector in disguise?
The second reason is I may want to introduce a `sort` function that can order the property entries according to the ordering of particles in the particle handler to ensure nice memory access during iteration over particles. I am not sure I need access to the available handles for that, but it may be useful. Anyway, the type is easy to change, if you insist on a stack and we decide we need something else later it is easy to change back.

Reply:
Nice. I agree to keep it this way for now as well.